docs/RFCS/20220818_tenant_version_upgrades.md

```diff
@@ -0,0 +1,475 @@
+   (`pkg/server/migration.go`, `BumpClusterVersion`)
+
+If the logical version needs to be upgraded across multiple steps (e.g. `v20.1-100` to `v20.1-200`),
+the logic in `Migrate` will step through the above 5 steps for every intermediate
+version: `v20.1-101`, `v20.1-102`, etc.
+
+It is the interlock between steps 2 and 5 that ensures the invariants are held.
+
+## Complementary+legacy mechanism: startup migrations
+
+The above section explains *cluster upgrade migrations* and how they
+are bound to logical versions.
+
+CockroachDB also contains a separate, older and legacy subsystem
+called *startup migrations*, which is not constrained by logical
+versions. (`pkg/startupmigrations`)
+
+This mechanism is simpler: the code inside each `cockroach` binary
+contains a list of *possible* startup migrations.
+
+Whenever a node starts up (regardless of logical version), it checks
+whether the startup migrations that it *can* run have already run. If
+they haven't, it runs them. The migrations are idempotent so that
+if two or more nodes start at the same time, it does not
+matter if they are running the same migration concurrently.
+
+Startup migrations pre-date the cluster upgrade migration subsystem
+described above, and is unable to respect logical version
+boundaries.
+
+We would prefer to use cluster upgrades for all its remaining uses,
+but this replacement was not done yet. (see issue
+https://github.com/cockroachdb/cockroach/issues/73813 ).
+
+Here are the remaining uses:
+
+- set the setting `diagnostics.reporting.enabled` to true unless the env var `COCKROACH_SKIP_ENABLING_DIAGNOSTIC_REPORTING` is set to false.
+- write an initial value to the `version` row in `system.settings`.
+- add a row for the `root` user in `system.users`.
+- add the `admin` role to `system.users` and make `root` a member of it in `system.role_members`.
+- auto-generate a random UUID for `cluster.secret`.
+- block the node startup if a user/role with name `public` is present in `system.users`.
+- create the `defaultdb` and `postgres` empty databases.
+- copy the values from the old `timeseries.storage.10s_resolution_ttl` and `timeseries.storage.30m_resolution_ttl` settings
+  to their new setting names.
+- add the default lat/long entries to `system.locations`.
+- add the `CREATELOGIN` option to roles that already have the `CREATEROLE` role option.
+
+## How these mechanisms have been extended with multi-tenancy
+
+### Startup migrations
+
+The simpler, legacy startup migration mechanism has a trivial extension to multi-tenancy:
+
+*Every time a SQL server for a secondary tenant starts, it attempts to run whichever
+startup migrations it can for its tenant system tables.*
+
+This is checked/done over and over again every time a tenant SQL server starts.
+
+It's also possible for different tenants to "step through" their
+startup migrations at different rates, and that's OK.
+
+### New concept: separate logical versions for tenants and the KV layer
```

Comment:
quip: this isn't "new": we already do this, and have documents and runbacks that assume this.